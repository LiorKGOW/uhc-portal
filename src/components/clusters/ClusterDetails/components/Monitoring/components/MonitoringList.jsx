import React from 'react';
import PropTypes from 'prop-types';
import { DataList } from '@patternfly/react-core';

import get from 'lodash/get';

import MonitoringListItem from './MonitoringListItem';
import AlertsTable from './AlertsTable';
import NodesTable from './NodesTable';
import ClusterOperators from './ClusterOperators';
import ResourceUsage from '../../../../common/ResourceUsage/ResourceUsage';
import MonitoringEmptyState from './MonitoringEmptyState';
import { metricsStatusMessages } from '../../../../common/ResourceUsage/ResourceUsage.consts';
import { hasResourceUsageMetrics } from '../monitoringHelper';
import { subscriptionStatuses } from '../../../../../../common/subscriptionTypes';

class MonitoringList extends React.Component {
    state = {
      expanded: [],
    }

  toggle = (sectionId) => {
    const { expanded } = this.state;
    const index = expanded.indexOf(sectionId);
    const newExpanded = index >= 0
      ? [...expanded.slice(0, index), ...expanded.slice(index + 1, expanded.length)]
      : [...expanded, sectionId];
    this.setState(() => ({ expanded: newExpanded }));
  };

  render() {
    const {
      cluster, alerts, nodes, operators, resourceUsage,
    } = this.props;
    const { expanded } = this.state;

    const metricsAvailable = hasResourceUsageMetrics(cluster);
    const isArchived = get(cluster, 'subscription.status', false) === subscriptionStatuses.ARCHIVED;
    const metricsStatusMessage = isArchived ? metricsStatusMessages.archived
      : metricsStatusMessages[cluster.state] || metricsStatusMessages.default;

    const EmptyState = (
      <MonitoringEmptyState hideLastCheckIn hideIcon title="No data available for this metric">
        <p>Please check back later</p>
      </MonitoringEmptyState>
    );

    return (
      <DataList aria-label="monitoring-table">
        <MonitoringListItem
          title="Alerts firing"
          numOfIssues={alerts.numOfIssues}
          numOfWarnings={alerts.numOfWarnings}
          toggle={this.toggle}
          expanded={expanded}
          hasData={alerts.hasData}
        >
          {alerts.hasData
            ? <AlertsTable alerts={alerts.data} clusterConsole={cluster.console} /> : EmptyState}
        </MonitoringListItem>
        <MonitoringListItem
          title="Nodes"
          numOfIssues={nodes.numOfIssues}
          toggle={this.toggle}
          expanded={expanded}
          hasData={nodes.hasData}
        >
          {nodes.hasData
            ? <NodesTable nodes={nodes.data} clusterConsole={cluster.console} /> : EmptyState }
        </MonitoringListItem>
        <MonitoringListItem
          title="Cluster operators"
          numOfIssues={operators.numOfIssues}
          numOfWarnings={operators.numOfWarnings}
          toggle={this.toggle}
          expanded={expanded}
          hasData={operators.hasData}
        >
          {operators.hasData
            ? <ClusterOperators operators={operators.data} clusterConsole={cluster.console} />
            : EmptyState }
        </MonitoringListItem>
        <MonitoringListItem
          title="Resource usage"
          numOfIssues={resourceUsage.numOfIssues}
          numOfWarnings={resourceUsage.numOfWarnings}
          toggle={this.toggle}
          expanded={expanded}
          hasData={resourceUsage.hasData}
        >
          {resourceUsage.hasData
            ? (
              <div className="metrics-chart">
                <ResourceUsage
                  cpu={{
                    used: cluster.metrics.cpu.used,
                    total: cluster.metrics.cpu.total,
                  }}
                  memory={{
                    used: cluster.metrics.memory.used,
                    total: cluster.metrics.memory.total,
                  }}
                  metricsAvailable={metricsAvailable}
                  metricsStatusMessage={metricsStatusMessage}
                  type="threshold"
                />
              </div>
            ) : EmptyState}
        </MonitoringListItem>
      </DataList>
    );
  }
}

MonitoringList.propTypes = {
  cluster: PropTypes.object,
  alerts: PropTypes.object,
  nodes: PropTypes.object,
  resourceUsage: PropTypes.object,
  operators: PropTypes.object,
};

MonitoringList.defaultProps = {
  cluster: {},
  alerts: {},
  nodes: {},
  resourceUsage: {},
  operators: {},
};

export default MonitoringList;
