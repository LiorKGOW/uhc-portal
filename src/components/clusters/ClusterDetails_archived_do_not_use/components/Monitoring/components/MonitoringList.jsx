import React from 'react';
import get from 'lodash/get';
import PropTypes from 'prop-types';

import { DataList } from '@patternfly/react-core';

import { SubscriptionCommonFields } from '~/types/accounts_mgmt.v1';

import { metricsStatusMessages } from '../../../../common/ResourceUsage/constants';
import ResourceUsage from '../../../../common/ResourceUsage/ResourceUsage';
import { hasResourceUsageMetrics } from '../monitoringHelper';

import AlertsTable from './AlertsTable';
import ClusterOperators from './ClusterOperators';
import MonitoringEmptyState from './MonitoringEmptyState';
import MonitoringListItem from './MonitoringListItem';
import NodesTable from './NodesTable';

class MonitoringList extends React.Component {
  state = {
    expanded: [],
  };

  toggle = (sectionId) => {
    const { expanded } = this.state;
    const index = expanded.indexOf(sectionId);
    const newExpanded =
      index >= 0
        ? [...expanded.slice(0, index), ...expanded.slice(index + 1, expanded.length)]
        : [...expanded, sectionId];
    this.setState(() => ({ expanded: newExpanded }));
  };

  render() {
    const { cluster, alerts, nodes, operators, resourceUsage } = this.props;
    const { expanded } = this.state;

    const metricsAvailable = hasResourceUsageMetrics(cluster);
    const isArchived =
      get(cluster, 'subscription.status', false) === SubscriptionCommonFields.status.ARCHIVED;
    const metricsStatusMessage = isArchived
      ? metricsStatusMessages.archived
      : metricsStatusMessages[cluster.state] || metricsStatusMessages.default;

    const EmptyState = (
      <MonitoringEmptyState hideLastCheckIn hideIcon title="No data available for this metric">
        <p>Please check back later</p>
      </MonitoringEmptyState>
    );

    return (
      <DataList aria-label="monitoring-table">
        <MonitoringListItem
          title="Alerts firing"
          numOfIssues={alerts.numOfIssues}
          numOfWarnings={alerts.numOfWarnings}
          toggle={this.toggle}
          expanded={expanded}
          hasData={alerts.hasData}
        >
          {alerts.hasData ? (
            <AlertsTable alerts={alerts.data} clusterConsole={cluster.console} />
          ) : (
            EmptyState
          )}
        </MonitoringListItem>
        {nodes.hasData && (
          // hide nodes if not available, this metric is missing in OpenShift >= 4.3.8
          <MonitoringListItem
            title="Nodes"
            numOfIssues={nodes.numOfIssues}
            toggle={this.toggle}
            expanded={expanded}
            hasData={nodes.hasData}
          >
            <NodesTable nodes={nodes.data} clusterConsole={cluster.console} />
          </MonitoringListItem>
        )}
        <MonitoringListItem
          title="Cluster operators"
          numOfIssues={operators.numOfIssues}
          numOfWarnings={operators.numOfWarnings}
          toggle={this.toggle}
          expanded={expanded}
          hasData={operators.hasData}
        >
          {operators.hasData ? (
            <ClusterOperators operators={operators.data} clusterConsole={cluster.console} />
          ) : (
            EmptyState
          )}
        </MonitoringListItem>
        <MonitoringListItem
          title="Resource usage"
          numOfIssues={resourceUsage.numOfIssues}
          numOfWarnings={resourceUsage.numOfWarnings}
          toggle={this.toggle}
          expanded={expanded}
          hasData={resourceUsage.hasData}
        >
          {resourceUsage.hasData ? (
            <div className="metrics-chart">
              <ResourceUsage
                cpu={{
                  used: cluster.metrics.cpu.used,
                  total: cluster.metrics.cpu.total,
                }}
                memory={{
                  used: cluster.metrics.memory.used,
                  total: cluster.metrics.memory.total,
                }}
                metricsAvailable={metricsAvailable}
                metricsStatusMessage={metricsStatusMessage}
                type="threshold"
              />
            </div>
          ) : (
            EmptyState
          )}
        </MonitoringListItem>
      </DataList>
    );
  }
}

MonitoringList.propTypes = {
  cluster: PropTypes.object,
  alerts: PropTypes.object,
  nodes: PropTypes.object,
  resourceUsage: PropTypes.object,
  operators: PropTypes.object,
};

MonitoringList.defaultProps = {
  cluster: {},
  alerts: {},
  nodes: {},
  resourceUsage: {},
  operators: {},
};

export default MonitoringList;
