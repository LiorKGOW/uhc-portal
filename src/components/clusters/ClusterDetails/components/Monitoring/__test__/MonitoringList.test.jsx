import React from 'react';
import { shallow } from 'enzyme';

import MonitoringList from '../components/MonitoringList';

describe('<MonitoringList />', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = shallow(<MonitoringList />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
