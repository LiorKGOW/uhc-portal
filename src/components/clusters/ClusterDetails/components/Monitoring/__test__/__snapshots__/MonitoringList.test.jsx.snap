// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MonitoringList /> should render 1`] = `
<DataList
  aria-label="monitoring-table"
>
  <MonitoringListItem
    expanded={Array []}
    title="Alerts firing"
    toggle={[Function]}
  >
    <MonitoringEmptyState
      hideIcon={true}
      hideLastCheckIn={true}
    >
      <p>
        Please check back later
      </p>
    </MonitoringEmptyState>
  </MonitoringListItem>
  <MonitoringListItem
    expanded={Array []}
    title="Nodes"
    toggle={[Function]}
  >
    <MonitoringEmptyState
      hideIcon={true}
      hideLastCheckIn={true}
    >
      <p>
        Please check back later
      </p>
    </MonitoringEmptyState>
  </MonitoringListItem>
  <MonitoringListItem
    expanded={Array []}
    title="Cluster operators"
    toggle={[Function]}
  >
    <MonitoringEmptyState
      hideIcon={true}
      hideLastCheckIn={true}
    >
      <p>
        Please check back later
      </p>
    </MonitoringEmptyState>
  </MonitoringListItem>
  <MonitoringListItem
    expanded={Array []}
    title="Resource usage"
    toggle={[Function]}
  >
    <MonitoringEmptyState
      hideIcon={true}
      hideLastCheckIn={true}
    >
      <p>
        Please check back later
      </p>
    </MonitoringEmptyState>
  </MonitoringListItem>
</DataList>
`;
